libvirt/uri/ssh.go

```diff
@@ -91,74 +125,184 @@ func (u *ConnectionURI) dialSSH() (net.Conn, error) {
-			u, err := user.Current()
+		hostKeyCallback = func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			err := kh(net.JoinHostPort(hostName, port), remote, key)
 			if err != nil {
-				return nil, fmt.Errorf("unable to get username: %w", err)
+				log.Printf("Host key verification failed for host '%s' (%s) %v: %v", hostName, remote, key, err)
 			}
-			sshu = u.Username
+			return err
 		}
-		username = sshu
+
+		keyAlgs, err := sshcfg.Get(target, "HostKeyAlgorithms")
+		if err == nil && keyAlgs != "" {
+			log.Printf("Got host key algorithms '%s'", keyAlgs)
+			hostKeyAlgorithms = strings.Split(keyAlgs, ",")
+		}
+
 	}
 
 	cfg := ssh.ClientConfig{
-		User:            username,
+		User:            u.User.Username(),
 		HostKeyCallback: hostKeyCallback,
-		Auth:            authMethods,
+		HostKeyAlgorithms: hostKeyAlgorithms,
 		Timeout:         dialTimeout,
 	}
 
-	port := u.Port()
-	if port == "" {
-		port = defaultSSHPort
+	proxy, err := sshcfg.Get(target, "ProxyCommand")
+	if err == nil && proxy != "" {
+		log.Printf("[WARNING] unsupported ssh ProxyCommand '%v'", proxy)
 	}
 
-	sshClient, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", u.Hostname(), port), &cfg)
-	if err != nil {
-		return nil, err
+	proxy, err = sshcfg.Get(target, "ProxyJump")
+	var bastion *ssh.Client
+	if err == nil && proxy != "" {
+		log.Printf("[DEBUG] found ProxyJump '%v'", proxy)
+
+		// this is a proxy jump: we recurse into that proxy
+		bastion, err = u.dialHost(proxy, sshcfg, depth + 1)
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to bastion host '%v': %w", proxy, err)
+		}
 	}
 
-	address := q.Get("socket")
-	if address == "" {
-		address = defaultUnixSock
+	if cfg.User == "" {
+		sshu, err := sshcfg.Get(target, "User")
+		log.Printf("[DEBUG] SSH User for target '%v' is '%v'", target, sshu)
+		if err != nil {
+			log.Printf("[DEBUG] ssh user: using current login")
+			u, err := user.Current()
+			if err != nil {
+				return nil, fmt.Errorf("unable to get username: %w", err)
+			}
+			sshu = u.Username
+		}
+		cfg.User = sshu
 	}
 
-	c, err := sshClient.Dial("unix", address)
-	if err != nil {
-		return nil, fmt.Errorf("failed to connect to libvirt on the remote host: %w", err)
+	cfg.Auth = u.parseAuthMethods(target, sshcfg)
+	if len(cfg.Auth) < 1 {
+		return nil, fmt.Errorf("could not configure SSH authentication methods")
 	}
 
-	return c, nil
+	if (bastion != nil) {
+		// if this is a proxied connection, we want to dial through the bastion host
+		log.Printf("[INFO] SSH connecting to '%v' (%v) through bastion host '%v'", target, hostName, proxy)
+		// Dial a connection to the service host, from the bastion
+		conn, err := bastion.Dial("tcp", net.JoinHostPort(hostName, port))
+		if err != nil {
+			log.Fatal(err)
+			return nil, fmt.Errorf("failed to connect to remote host '%v': %w", target, err)
+		}
+
+		ncc, chans, reqs, err := ssh.NewClientConn(conn, target, &cfg)
+		if err != nil {
+			log.Fatal(err)
```

Comment:
This makes no sense. `log.Fatal` will `os.Exit` so the next line will never be executed, and you don't want to use `os.Exit` in a plugin. Just return the error, unless you wanted to achieve something else.

Reply:
My bad. I was unfamiliar with go's log.Fatal and I assumed it was simply like in python indicating logging facility severity.

I have removed log.Fatal as the intent was simply to return a high severity log trace while returning the error up.